x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/model/tree/Tree.java

```diff
@@ -0,0 +1,311 @@
+            nodes.add(null);
+            addLeaf(0, 0.0);
+            numNodes = 1;
+        }
+
+        public Builder setFeatureNames(List<String> featureNames) {
+            this.featureNames = featureNames;
+            return this;
+        }
+
+        public Builder addNode(TreeNode.Builder node) {
+            nodes.add(node);
+            return this;
+        }
+
+        public Builder setNodes(List<TreeNode.Builder> nodes) {
+            this.nodes = new ArrayList<>(nodes);
+            return this;
+        }
+
+        public Builder setNodes(TreeNode.Builder... nodes) {
+            return setNodes(Arrays.asList(nodes));
+        }
+
+        /**
+         * Add a decision node. Space for the child nodes is allocated
+         * @param nodeIndex         Where to place the node. This is either 0 (root) or an existing child node index
+         * @param featureIndex      The feature index the decision is made on
+         * @param isDefaultLeft     Default left branch if the feature is missing
+         * @param decisionThreshold The decision threshold
+         * @return The created node
+         */
+        TreeNode.Builder addJunction(int nodeIndex, int featureIndex, boolean isDefaultLeft, double decisionThreshold) {
+            int leftChild = numNodes++;
+            int rightChild = numNodes++;
+            nodes.ensureCapacity(nodeIndex + 1);
+            for (int i = nodes.size(); i < nodeIndex + 1; i++) {
+                nodes.add(null);
+            }
+
+            TreeNode.Builder node = TreeNode.builder()
+                .setDefaultLeft(isDefaultLeft)
+                .setLeftChild(leftChild)
+                .setRightChild(rightChild)
+                .setSplitIndex(featureIndex).setThreshold(decisionThreshold);
+            nodes.set(nodeIndex, node);
+
+            // allocate space for the child nodes
+            while (nodes.size() <= rightChild) {
+                nodes.add(null);
+            }
+
+            return node;
+        }
+
+        void detectCycle(List<TreeNode.Builder> nodes) {
+            if (nodes.isEmpty()) {
+                return;
+            }
+            Set<Integer> visited = new HashSet<>();
+            Queue<Integer> toVisit = new ArrayDeque<>(nodes.size());
+            toVisit.add(0);
+            while(toVisit.isEmpty() == false) {
+                Integer nodeIdx = toVisit.remove();
+                if (visited.contains(nodeIdx)) {
+                    throw new IllegalArgumentException("[tree] contains cycle at node " + nodeIdx);
+                }
+                visited.add(nodeIdx);
+                TreeNode.Builder treeNode = nodes.get(nodeIdx);
+                if (treeNode.getLeftChild() != null) {
+                    toVisit.add(treeNode.getLeftChild());
+                }
+                if (treeNode.getRightChild() != null) {
+                    toVisit.add(treeNode.getRightChild());
+                }
+            }
```

Comment:
you could add an easy check for disconnected nodes by checking `visited.size() == nodes.size()`